mlflow/types/schema.py

```diff
@@ -406,3 +425,307 @@ def __eq__(self, other) -> bool:
 
     def __repr__(self) -> str:
         return repr(self.inputs)
+
+
+@experimental
+class ParamSpec:
+    """
+    Specification used to represent parameters for the model.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        type: Union[DataType, str],  # pylint: disable=redefined-builtin
+        default: Union[DataType, List[DataType], None],
+        shape: Optional[tuple] = None,
+    ):
+        self._name = str(name)
+        self._shape = tuple(shape) if shape is not None else None
+
+        try:
+            self._type = DataType[type] if isinstance(type, str) else type
+        except KeyError:
+            raise MlflowException(
+                f"Unsupported type '{type}', expected instance of DataType or "
+                f"one of {[t.name for t in DataType]}"
+            )
+        if not isinstance(self.type, DataType):
+            raise TypeError(
+                "Expected mlflow.models.signature.Datatype or str for the 'type' "
+                f"argument, but got {self.type.__class__}"
+            )
+
+        # This line makes sure repr(self) works fine
+        self._default = default
+        self._default = self.validate_type_and_shape(repr(self), default, self.type, self.shape)
+
+    @classmethod
+    def validate_param_spec(
+        cls, value: Union[DataType, List[DataType], None], param_spec: "ParamSpec"
+    ):
+        return cls.validate_type_and_shape(
+            repr(param_spec), value, param_spec.type, param_spec.shape
+        )
+
+    @classmethod
+    def enforce_param_datatype(cls, name, value, t: DataType):
+        """
+        Enforce the value matches the data type.
+
+        The following type conversions are allowed:
+
+        1. int -> long, float, double
+        2. long -> float, double
+        3. float -> double
+        4. any -> datetime (try conversion)
+        5. str -> bytes
+
+        Any other type mismatch will raise error.
+
+        :param name: parameter name
+        :param value: parameter value
+        :param t: expected data type
+        """
+        if value is None:
+            return value
+
+        if not np.isscalar(value):
+            raise MlflowException(
+                f"Value should be a scalar for param {name}, got {value}", INVALID_PARAMETER_VALUE
+            )
+
+        if DataType.is_instance(value, t):
+            return value
+
+        if isinstance(value, str) and t == DataType.binary:
+            try:
+                return base64.decodebytes(value.encode("ascii"))
```

Comment:
Should we encode with 'ascii' or 'utf-8' here? It might depends on user's input?

Reply:
https://github.com/mlflow/mlflow/blob/f0649dcca943facf0a81d5dd8f104ae06bec532c/mlflow/utils/proto_json_utils.py#L159 NumpyEncoder uses 'ascii' when it convert responses, so in order to make the response consistent I use 'ascii' here as well. But in this case, 'utf-8' encoded byte string is not recognizable

Reply:
`ascii` wont' work if `value` is like `value = "Hello, world! 👋"`.

Reply:
Maybe we should reject this type for parameters by the moment? Indeed, not sure if there is a use case for it. It is more likely to be a model input.

Reply:
@serena-ruan @harupy @santiagxf I agree, let's drop support for binary params for now.